File: AutoUpdate/AutoUpdate/Services/AutoUpdateService.cs
using AutoUpdate.Configuration;
using AutoUpdate.Models;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace AutoUpdate.Services;

/// <summary>
/// Main implementation of IAutoUpdateService that orchestrates the update process.
/// </summary>
public class AutoUpdateService : IAutoUpdateService
{
    private readonly AutoUpdateOptions _options;
    private readonly IVersionProvider _versionProvider;
    private readonly IUpdateDownloader _updateDownloader;
    private readonly IUserDataManager _userDataManager;
    private readonly ILogger<AutoUpdateService> _logger;

    public AutoUpdateService(
        IOptions<AutoUpdateOptions> options,
        IVersionProvider versionProvider,
        IUpdateDownloader updateDownloader,
        IUserDataManager userDataManager,
        ILogger<AutoUpdateService> logger)
    {
        _options = options.Value;
        _versionProvider = versionProvider;
        _updateDownloader = updateDownloader;
        _userDataManager = userDataManager;
        _logger = logger;
    }

    /// <inheritdoc/>
    public async Task<UpdateCheckResult> CheckForUpdatesAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            _logger.LogInformation("Checking for updates from GitHub repository: {Owner}/{Repo} branch: {Branch}",
                _options.GitHubOwner, _options.GitHubRepo, _options.ProductionBranch);

            // Get local and remote versions
            var localVersion = await _versionProvider.GetLocalVersionAsync(cancellationToken);
            var remoteVersion = await _versionProvider.GetRemoteVersionAsync(
                _options.GitHubOwner,
                _options.GitHubRepo,
                _options.ProductionBranch,
                cancellationToken);

            if (remoteVersion == null)
            {
                _logger.LogWarning("Failed to fetch remote version information");
                return UpdateCheckResult.NoUpdate(localVersion);
            }

            // Compare versions
            if (remoteVersion.IsGreaterThan(localVersion))
            {
                _logger.LogInformation("Update available: {Current} -> {New}",
                    localVersion.Version, remoteVersion.Version);
                return UpdateCheckResult.UpdateAvailable(localVersion, remoteVersion);
            }
            else
            {
                _logger.LogInformation("Application is up to date: {Version}", localVersion.Version);
                return UpdateCheckResult.NoUpdate(localVersion);
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error checking for updates");
            // Return no update available on error
            var localVersion = await _versionProvider.GetLocalVersionAsync(cancellationToken);
            return UpdateCheckResult.NoUpdate(localVersion);
        }
    }

    /// <inheritdoc/>
    public async Task<UpdateResult> ApplyUpdateAsync(CancellationToken cancellationToken = default)
    {
        var tempPath = Path.Combine(Path.GetTempPath(), $"AutoUpdate_{Guid.NewGuid()}");

        try
        {
            _logger.LogInformation("Starting update process");

            // 1. Get remote version info
            var remoteVersion = await _versionProvider.GetRemoteVersionAsync(
                _options.GitHubOwner,
                _options.GitHubRepo,
                _options.ProductionBranch,
                cancellationToken);

            if (remoteVersion == null)
            {
                return UpdateResult.Failure("Failed to fetch remote version information");
            }

            // 2. Backup user data
            _logger.LogInformation("Backing up user data");
            await _userDataManager.BackupAsync(_options.UserDataPaths, _options.BackupDirectory, cancellationToken);

            // 3. Download and extract update
            _logger.LogInformation("Downloading and extracting update");
            var extractedPath = await _updateDownloader.DownloadAndExtractAsync(
                _options.GitHubOwner,
                _options.GitHubRepo,
                _options.ProductionBranch,
                tempPath,
                cancellationToken);

            // 4. Apply update files (exclude user data and version file)
            var excludePaths = _options.UserDataPaths.Concat(new[] { _options.VersionFilePath }).ToArray();
            _logger.LogInformation("Applying update files");
            await _updateDownloader.ApplyFilesAsync(extractedPath, excludePaths, cancellationToken);

            // 5. Update local version file
            _logger.LogInformation("Updating version file to: {Version}", remoteVersion.Version);
            var versionJson = System.Text.Json.JsonSerializer.Serialize(remoteVersion, new System.Text.Json.JsonSerializerOptions
            {
                WriteIndented = true
            });
            await File.WriteAllTextAsync(_options.VersionFilePath, versionJson, cancellationToken);

            // 6. Restore user data
            _logger.LogInformation("Restoring user data");
            await _userDataManager.RestoreAsync(_options.BackupDirectory, _options.UserDataPaths, cancellationToken);

            _logger.LogInformation("Update completed successfully: {Version}", remoteVersion.Version);
            return UpdateResult.Success(remoteVersion);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Update failed, attempting to restore user data");

            // Try to restore user data on failure
            try
            {
                await _userDataManager.RestoreAsync(_options.BackupDirectory, _options.UserDataPaths, cancellationToken);
                _logger.LogInformation("User data restored after failed update");
            }
            catch (Exception restoreEx)
            {
                _logger.LogError(restoreEx, "Failed to restore user data after update failure");
            }

            return UpdateResult.Failure($"Update failed: {ex.Message}");
        }
        finally
        {
            // Clean up temporary files
            try
            {
                if (Directory.Exists(tempPath))
                {
                    Directory.Delete(tempPath, true);
                    _logger.LogDebug("Temporary files cleaned up");
                }
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to clean up temporary files");
            }
        }
    }
}

File: AutoUpdateJS/src/services/AutoUpdateService.js
import { UpdateCheckResult } from '../models/UpdateCheckResult.js';
import { UpdateResult } from '../models/UpdateResult.js';
import { VersionProvider } from './VersionProvider.js';
import { UpdateDownloader } from './UpdateDownloader.js';
import { UserDataManager } from './UserDataManager.js';
import { writeJsonFile, createTempDirectory, remove } from '../utils/fileUtils.js';
import logger from '../utils/logger.js';

/**
 * Main service that orchestrates the auto-update process.
 */
export class AutoUpdateService {
  /**
   * Creates a new AutoUpdateService instance.
   * @param {AutoUpdateConfig} config - Configuration instance
   */
  constructor(config) {
    this.config = config;

    // Initialize services
    this.versionProvider = new VersionProvider(config);
    this.updateDownloader = new UpdateDownloader(config);
    this.userDataManager = new UserDataManager(config);

    // Configure logging
    logger.setLoggingEnabled(config.enableLogging);
  }

  /**
   * Checks for available updates.
   * @returns {Promise<UpdateCheckResult>} Update check result
   */
  async checkForUpdates() {
    try {
      logger.info(`Checking for updates from GitHub repository: ${this.config.githubOwner}/${this.config.githubRepo} branch: ${this.config.productionBranch}`);

      const updateInfo = await this.versionProvider.checkForUpdate();

      if (updateInfo.hasUpdate) {
        logger.info(`Update available: ${updateInfo.localVersion.version} -> ${updateInfo.remoteVersion.version}`);
        return UpdateCheckResult.updateAvailable(updateInfo.localVersion, updateInfo.remoteVersion);
      } else {
        logger.info(`Application is up to date: ${updateInfo.localVersion.version}`);
        return UpdateCheckResult.noUpdate(updateInfo.localVersion);
      }
    } catch (error) {
      logger.error(`Error checking for updates: ${error.message}`, error);

      // Return no update on error
      const localVersion = await this.versionProvider.getLocalVersion();
      return UpdateCheckResult.noUpdate(localVersion);
    }
  }

  /**
   * Applies available updates.
   * @returns {Promise<UpdateResult>} Update result
   */
  async applyUpdate() {
    let tempPath = null;

    try {
      logger.info('Starting update process');

      // 1. Get remote version info
      const remoteVersion = await this.versionProvider.getProductionVersion();
      if (!remoteVersion) {
        const errorMsg = 'Failed to fetch remote version information';
        logger.error(errorMsg);
        return UpdateResult.failure(errorMsg);
      }

      // 2. Backup user data
      logger.info('Backing up user data');
      await this.userDataManager.backupUserData();

      // 3. Create temp directory for update process
      tempPath = await createTempDirectory('autoupdate');

      // 4. Download and extract update
      logger.info('Downloading and extracting update');
      const extractedPath = await this.updateDownloader.downloadAndExtract(
        this.config.githubOwner,
        this.config.githubRepo,
        this.config.productionBranch,
        tempPath
      );

      // 5. Apply update files (exclude user data and version file)
      const excludePaths = [
        ...this.config.userDataPaths,
        this.config.versionFilePath
      ];

      logger.info('Applying update files');
      await this.updateDownloader.applyFiles(extractedPath, excludePaths);

      // 6. Update local version file
      logger.info(`Updating version file to: ${remoteVersion.version}`);
      await writeJsonFile(this.config.versionFilePath, remoteVersion.toJSON());

      // 7. Restore user data
      logger.info('Restoring user data');
      await this.userDataManager.restoreUserData();

      logger.info(`Update completed successfully: ${remoteVersion.version}`);
      return UpdateResult.success(remoteVersion);

    } catch (error) {
      logger.error(`Update failed: ${error.message}`, error);

      // Try to restore user data on failure
      try {
        logger.info('Attempting to restore user data after failed update');
        await this.userDataManager.restoreUserData();
      } catch (restoreError) {
        logger.error(`Failed to restore user data: ${restoreError.message}`, restoreError);
      }

      return UpdateResult.failure(`Update failed: ${error.message}`);
    } finally {
      // Clean up temporary files
      if (tempPath) {
        try {
          await remove(tempPath);
          logger.debug('Temporary files cleaned up');
        } catch (error) {
          logger.debug(`Failed to clean up temporary files: ${error.message}`);
        }
      }
    }
  }

  /**
   * Performs a complete update check and apply cycle.
   * @returns {Promise<{checkResult: UpdateCheckResult, updateResult: UpdateResult|null}>}
   */
  async checkAndApplyUpdate() {
    const checkResult = await this.checkForUpdates();

    if (!checkResult.hasUpdate) {
      return { checkResult, updateResult: null };
    }

    const updateResult = await this.applyUpdate();
    return { checkResult, updateResult };
  }

  /**
   * Gets the current local version.
   * @returns {Promise<VersionInfo>} Current version
   */
  async getCurrentVersion() {
    return this.versionProvider.getLocalVersion();
  }

  /**
   * Gets the latest remote version.
   * @returns {Promise<VersionInfo|null>} Latest remote version
   */
  async getLatestVersion() {
    return this.versionProvider.getProductionVersion();
  }

  /**
   * Cleans up old backup files.
   * @param {number} keepCount - Number of backups to keep
   * @returns {Promise<void>}
   */
  async cleanupBackups(keepCount = 5) {
    await this.userDataManager.cleanupOldBackups(keepCount);
  }

  /**
   * Validates the current configuration.
   * @returns {string[]} Array of validation errors (empty if valid)
   */
  validateConfig() {
    return this.config.validate();
  }
}
